radix-engine/src/engine/kernel.rs

```diff
@@ -949,7 +970,10 @@ where
 
         // TODO: Move this into higher layer, e.g. transaction processor
         if self.current_frame.depth == 0 {
-            self.current_frame.drop_frame(&mut self.heap)?;
+            self.current_frame.drop_all_locks(&mut self.track)?;
+            self.execution_mode = ExecutionMode::Application;
+            self.drop_nodes_in_frame()?;
+            self.execution_mode = ExecutionMode::Kernel;
```

Comment:
I think the current of execution mode management is very fragile and error prone. Needs a better machanism for execution mode switching. :)

Reply:
Yup, I didn't totally finish this part, just wanted to get the PR out as soon as possible :) will cleanup :)